osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -202,6 +204,9 @@ public async Task ChangeState(MultiplayerUserState newState)
                 if (user == null)
                     failWithInvalidState("Local user was not found in the expected room");
 
+                if (user.BeatmapAvailability.State != DownloadState.LocallyAvailable)
+                    throw new InvalidStateException("Attempted to change user state while they don't have the room's beatmap.");
```

Comment:
is wrong. will possibly throw if host changes beatmap while another user is still at results and keep the user stuck at results, depending on client implementation. this should only prevent changes to anything other than `Idle` if the map is not downloaded.

Reply:
this won't be the case, `ChangeBeatmapAvailability` will force the user state back to `Idle`, and clients probably should handle that change and take the user out of results screen and into the room sub screen.

Reply:
the client is responsible for sending beatmap availability. will it be sent from the client in results screen if the host changes map? because if not, then this will throw. which is why i said "depending on client implementation".

also disagree on the latter part of that. if the user wants to keep viewing results, they should be free to do so for however long they choose.

Reply:
so just to be sure, you're fine with a user that is in `Ready` state, when host changes beatmap user doesn't have, the user will not get forced back to `Idle`, but let that be handled by the client instead, right?

Reply:
i'm confused by this question. the server can't know if the user has the beatmap or not. unless you want to explicitly unready every user in the room on a map change. in fact, that might actually be safer.

Reply:
hmm, actually ok, can change the `Idle` revert to client, was thinking of the availability becoming more of something a server gets and handle whoever user has it properly, but it's starting to look like it should be more of just an information.

Reply:
on second thought maybe this isn't as much of an issue as any setting change in the room reverts all users back to `Idle` already - but i'd still permit changing state to `Idle` regardless of whether you have the map or not, as it doesn't do any harm and prevents potential errors

Reply:
yeah sure.